### PR TITLE
feat: integrate rss3 features

### PR DESCRIPTION
**Pull request by UncleBill**
successor of #4921

closes #4921

### CODE REVIEW

File: packages/mask/src/plugins/Profile/SNSAdaptor/ConnectRSS3.tsx [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                                    onClick={oneKeyActivate}
+                                    fontSize="text-md"
+                                    width="w-60"
+                                    height="h-12"
+                                    text="One Key Registration"
+                                />
+                                <div>
+                                    <Typography className={classes.msg} variant="body2" gutterBottom>
+                                        By clicking the button above, you agree to the{' '}
+                                        <Link href="https://rss3.io/#privacy" target="_blank">
```
Comment (septs) [inline packages/mask/src/plugins/Profile/SNSAdaptor/ConnectRSS3.tsx:112]:
```suggestion
                                        <Link href="https://rss3.io#privacy" target="_blank">
```

File: packages/mask/src/plugins/Profile/SNSAdaptor/hooks/useFootprints.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { useEffect, useState } from 'react'
+import type { GeneralAssetWithTags } from '../common/types'
+import utils from '../common/utils'
+
+export function useFootprints() {
+    const [footprints, setFootprints] = useState<GeneralAssetWithTags[]>([])
+    const [isLoading, setIsLoading] = useState(true)
+
+    useEffect(() => {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/SNSAdaptor/hooks/useFootprints.ts:9]: Refactor to use [`useAsync`](https://github.com/streamich/react-use/blob/master/docs/useAsync.md).

File: packages/mask/src/plugins/Profile/SNSAdaptor/hooks/useDonations.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { useEffect, useState } from 'react'
+import type { GeneralAssetWithTags } from '../common/types'
+import utils from '../common/utils'
+
+export function useDonations() {
+    const [donations, setDonations] = useState<GeneralAssetWithTags[]>([])
+    const [isLoading, setIsLoading] = useState(true)
+
+    useEffect(() => {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/SNSAdaptor/hooks/useDonations.ts:9]: Refactor to use [`useAsync`](https://github.com/streamich/react-use/blob/master/docs/useAsync.md).

File: packages/mask/src/plugins/Profile/SNSAdaptor/components/FootprintCard.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    startDate: string | undefined
+    endDate: string | undefined
+    city: string | undefined
+    country: string | undefined
+    username: string
+    activity: string
+}
+
+const formatDate = (ts: string): string => {
+    return new Date(Number.parseInt(ts, 16) * 1000).toLocaleDateString('en-US')
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/SNSAdaptor/components/FootprintCard.tsx:16]: import fromUnixTime from 'date-fns/fromUnixTime'

File: packages/mask/src/plugins/Profile/SNSAdaptor/common/rss3.ts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+function dispatchEvent(event: string, detail: any) {
+    const evt = new CustomEvent(event, {
+        detail,
+        bubbles: true,
+        composed: true,
+    })
+    document.dispatchEvent(evt)
+    console.log(event, detail)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/SNSAdaptor/common/rss3.ts:79]:
```suggestion
```

File: packages/mask/src/plugins/Profile/SNSAdaptor/common/address.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+const formatter = (address: string | undefined): string => {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/SNSAdaptor/common/address.ts:1]: It duplicated `fromatEthereumAddress`.

Comment (UncleBill) [reply packages/mask/src/plugins/Profile/SNSAdaptor/common/address.ts:1]: It's unused. Removed

**Comment by wenluomask** [comment]
1. Since I just recovered my backup but didn't connect any wallet, it supposed to get an address from NFT avatar and display that address' NFTs. In this case, Web3 tab didn't display anything.(I tried to add an ens in bio and then delete later, Web3 can fetch avatar's address);
Screen Shot 2021-12-19 at 10 47 03 PM

2. Only ens can relate to RSS3 account. If I used address that registered in RSS3, it didn't recognize it is registered in RSS3. The address should be the relation of Mask, Twitter and RSS3, but not ens;

3. The current display of address should be the same format in three tabs;

4. "Contribution" cannot display fully, hopefully you can adjust the UI(Although the UI is as same as RSS3);
Screen Shot 2021-12-19 at 11 26 35 PM

5. Attended should be capital;
Screen Shot 2021-12-19 at 11 31 58 PM

**Comment by UncleBill** [comment]
> 1. Since I just recovered my backup but didn't connect any wallet, it supposed to get an address from NFT avatar and display that address' NFTs. In this case, Web3 tab didn't display anything.(I tried to add an ens in bio and then delete later, Web3 can fetch avatar's address);

> 
> Screen Shot 2021-12-19 at 10 47 03 PM
> 
> 2. Only ens can relate to RSS3 account. If I used address that registered in RSS3, it didn't recognize it is registered in RSS3. The address should be the relation of Mask, Twitter and RSS3, but not ens;

Now we can resolve use address from rss3 profile link

> 3. The current display of address should be the same format in three tabs;
Fixed
> 4. "Contribution" cannot display fully, hopefully you can adjust the UI(Although the UI is as same as RSS3);

> 
> Screen Shot 2021-12-19 at 11 26 35 PM
> 

"Contrib" is not a typo, you can check out https://joshua.rss3.bio/gitcoins

> 5. Attended should be capital;
> 
> Screen Shot 2021-12-19 at 11 31 58 PM

Fixed

File: packages/mask/src/plugins/Profile/services.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        type: 'Gitcoin-Donation',
+    })
+
+    return fetchJSON(url) as Promise<Response>
+}
+
+export function getFootprints(account: string): Promise<Response> {
+    const url = urlcat('https://hub.pass3.me/assets/list', {
+        personaID: account,
+        type: 'POAP',
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/services.ts:34]: An enum for this.

```ts
enum AssetType {
    NFT = 'NFT',
    GitcoinDonation = 'Gitcoin-Donation',
    POAP = 'POAP',
}
```

File: packages/mask/src/plugins/Profile/apis/index.ts [hunk 8]
```diff
@@ -0,0 +1 @@
+export * from './rss3'
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/apis/index.ts:1]: Remove this file.

**Comment by guanbinrui** [comment]
@UncleBill Please remove all of the unused files.

File: packages/mask/src/plugins/Profile/SNSAdaptor/DonationsPage.tsx [hunk 9]
```diff
@@ -1,11 +1,74 @@
...
+                {donations.map((donation) => (
+                    <Link
+                        className={classes.link}
+                        href={getDonationLink(address, donation)}
+                        key={donation.id}
+                        target="_blank"
+                        rel="noopener noreferrer">
+                        <DonationCard
+                            imageUrl={donation.info.image_preview_url || RSS3_DEFAULT_IMAGE}
+                            name={donation.info.title || 'Inactive Project'}
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/SNSAdaptor/DonationsPage.tsx:65]: i18n

File: packages/mask/src/plugins/Profile/services.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    profile: RSS3Profile
+}
+
+export function getDonations(account: string): Promise<Response> {
+    const url = urlcat('https://hub.pass3.me/assets/list', {
+        personaID: account,
+        type: AssetType.GitcoinDonation,
+    })
+
+    return fetchJSON(url) as Promise<Response>
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/services.ts:26]:
```suggestion
    return fetchJSON<Response>(url)
```

File: packages/mask/src/plugins/Profile/services.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+export function getFootprints(account: string): Promise<Response> {
+    const url = urlcat('https://hub.pass3.me/assets/list', {
+        personaID: account,
+        type: AssetType.POAP,
+    })
+
+    return fetchJSON(url) as Promise<Response>
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/services.ts:35]:
```suggestion
    return fetchJSON<Response>(url)
```

File: packages/mask/src/plugins/Profile/services.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+export async function getAddressByRss3ProfileLink(profileUrl: string) {
+    const matched = profileUrl.match(/^https?:\/\/(\w+)\.rss3\.bio/)
+    if (!matched) {
+        return ''
+    }
+    const url = urlcat('https://rss3.domains/name/:id', { id: matched[1] })
+    const rsp = (await fetchJSON(url)) as NameInfo
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/services.ts:44]:
```suggestion
    const rsp = (await fetchJSON<NameInfo>(url))
```

File: packages/mask/src/plugins/Profile/services.ts [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    return rsp.address
+}
+
+export async function getRss3Profile(address: string) {
+    const url = urlcat('https://hub.pass3.me/:address', { address })
+    const rsp = (await fetchJSON(url)) as RSS3Info
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/services.ts:50]:
```suggestion
    const rsp = (await fetchJSON<RSS3Info>(url))
```

File: packages/mask/src/plugins/Profile/SNSAdaptor/hooks/useAddressByRss3ProfileLink.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
+import { useAsync } from 'react-use'
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/SNSAdaptor/hooks/useAddressByRss3ProfileLink.ts:1]: Rename this file by the name of the hook in it.

File: packages/mask/src/plugins/Profile/SNSAdaptor/hooks/useAddressByRss3ProfileLink.ts [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+export function useAddressByRss3(profileUrl: string, nickname: string) {
+    const { value: address, loading } = useAsync(async () => {
+        let rss3id: string
+        let matched = nickname.match(RSS3_RNS_RE)
+        if (matched) {
+            rss3id = matched[1]
+        } else {
+            matched = profileUrl.match(RSS3_PROFILE_URL_RE)
+            rss3id = matched ? matched[1] : ''
+        }
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Profile/SNSAdaptor/hooks/useAddressByRss3ProfileLink.ts:14]:
```suggestion
        const matched = nickname.match(RSS3_RNS_RE) || profileUrl.match(RSS3_PROFILE_URL_RE)
        const id = matched ? matched[1] : ''
```

File: packages/mask/src/plugins/Profile/constants.ts [hunk 16]
```diff
@@ -3,3 +3,7 @@ export const PLUGIN_NAME = 'Profile'
 export const PLUGIN_DESCRIPTION = 'Profile'
 
 export const RSS3_DEFAULT_IMAGE = 'https://rss3.mypinata.cloud/ipfs/QmVFq9qimnudPcs6QkQv8ZVEsvwD3aqETHWtS5yXgdbYY5'
+
+export const RSS3_PROFILE_URL_RE = /^https?:\/\/(\w+)\.rss3\.bio/
```
Comment (septs) [inline packages/mask/src/plugins/Profile/constants.ts:7]: suggest, use named groups?

```suggestion
export const RSS3_PROFILE_URL_RE = /^https?:\/\/(?<name>\w+)\.rss3\.bio/
```

```
> pattern = /^https?:\/\/(?<name>\w+)\.rss3\.bio/
/^https?:\/\/(?<name>\w+)\.rss3\.bio/
> matched = pattern.exec('https://cz_binance.rss3.bio/')
[
  'https://cz_binance.rss3.bio',
  'cz_binance',
  index: 0,
  input: 'https://cz_binance.rss3.bio/',
  groups: [Object: null prototype] { name: 'cz_binance' }
]
> matched.groups.name
'cz_binance'
```

File: packages/mask/src/plugins/Profile/constants.ts [hunk 17]
```diff
@@ -1,3 +1,9 @@
 export const PLUGIN_ID = 'com.maskbook.profile'
 export const PLUGIN_NAME = 'Profile'
 export const PLUGIN_DESCRIPTION = 'Profile'
+
+export const RSS3_DEFAULT_IMAGE = 'https://rss3.mypinata.cloud/ipfs/QmVFq9qimnudPcs6QkQv8ZVEsvwD3aqETHWtS5yXgdbYY5'
+
+export const RSS3_PROFILE_URL_RE = /^https?:\/\/(\w+)\.rss3\.bio/
+
+export const RSS3_RNS_RE = /^(\w+)\.rss3$/
```
Comment (septs) [inline packages/mask/src/plugins/Profile/constants.ts:9]:
```suggestion
export const RSS3_RNS_RE = /^(?<name>\w+)\.rss3$/
```

File: packages/mask/src/plugins/Profile/constants.ts (hunk 16 above)
Comment (UncleBill) [reply packages/mask/src/plugins/Profile/constants.ts:7]: iOS doesn't support regexp group yet.

File: packages/mask/src/plugins/Profile/constants.ts (hunk 17 above)
Comment (UncleBill) [reply packages/mask/src/plugins/Profile/constants.ts:9]: ditto

File: packages/mask/src/plugins/Profile/constants.ts (hunk 16 above)
Comment (septs) [reply packages/mask/src/plugins/Profile/constants.ts:7]: > iOS doesn't support regexp group yet.

Safari 15.2 supported

Comment (Jack-Works) [reply packages/mask/src/plugins/Profile/constants.ts:7]: image

All green according to https://kangax.github.io/compat-table/es2016plus/

File: packages/mask/background/services/helper/fetch.ts [hunk 18]
```diff
@@ -3,7 +3,8 @@ export async function fetch(url: string) {
     return res.blob()
 }
 
-export async function fetchJSON(url: string): Promise<unknown> {
+export async function fetchJSON<T extends unknown>(url: string): Promise<T> {
```
Comment (septs) [inline packages/mask/background/services/helper/fetch.ts:6]: plz sync develop branch

**Comment by wenluomask** [comment]
https://mask.atlassian.net/jira/software/c/projects/MASK/boards/20?modal=detail&selectedIssue=MASK-314

File: packages/mask/src/plugins/Profile/SNSAdaptor/components/AddressViewer.tsx [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+        t('plugin_profile_binding_rule4'),
+    ]
+
+    const tooltipRender = (
+        <div style={{ textAlign: 'left' }}>
+            <Typography variant="body2">{t('plugin_profile_binding_rules_title')}</Typography>
+            <ul className={classes.tipList}>
+                {rulesTipMap.map((item, index) => {
+                    return <li key={index}>{item}</li>
+                })}
```
Comment (septs) [inline packages/mask/src/plugins/Profile/SNSAdaptor/components/AddressViewer.tsx:53]:
```suggestion
                {rulesTipMap.map((item, index) => <li key={index}>{item}</li>)}
```

File: packages/mask/src/plugins/Wallet/services/domain.ts [hunk 20]
```diff
@@ -1,22 +1,141 @@
...
+
+function getEthereumName(twitterId: string, nickname: string, bio: string) {
+    const [matched] = nickname.match(ENS_RE) ?? bio.match(ENS_RE) ?? []
+    if (matched) return matched
+    return twitterId && !twitterId.endsWith('.eth') ? `${twitterId}.eth` : ''
+}
+
+function getRSS3Id(nickname: string, profileURL: string) {
+    const matched = nickname.match(RSS3_RNS_RE) || profileURL.match(RSS3_PROFILE_URL_RE)
+    return matched ? matched.groups?.name ?? '' : ''
```
Comment (septs) [inline packages/mask/src/plugins/Wallet/services/domain.ts:30]:
```suggestion
    return matched?.groups?.name ?? ''
```

File: packages/mask/src/plugins/Wallet/services/domain.ts [hunk 21]
```diff
@@ -1,22 +1,141 @@
...
+const RSS3_RNS_RE = /^(?<name>\w+)\.rss3$/
+
+function isValidAddress(address: string) {
+    return address && EthereumAddress.isValid(address) && !isSameAddress(address, ZERO_ADDRESS)
+}
+
+function getEthereumName(twitterId: string, nickname: string, bio: string) {
+    const [matched] = nickname.match(ENS_RE) ?? bio.match(ENS_RE) ?? []
+    if (matched) return matched
+    return twitterId && !twitterId.endsWith('.eth') ? `${twitterId}.eth` : ''
```
Comment (septs) [inline packages/mask/src/plugins/Wallet/services/domain.ts:25]:
```suggestion
    return !twitterId?.endsWith('.eth') ? `${twitterId}.eth` : ''
```

File: packages/web3-shared/evm/types/index.ts [hunk 22]
```diff
@@ -609,15 +609,19 @@ export interface Transaction {
 
 //#region address name
 export enum AddressNameType {
-    ENS = 'ENS',
-    UNS = 'UNS',
-    DNS = 'DNS',
+    ADDRESS = 1,
```
Comment (septs) [inline packages/web3-shared/evm/types/index.ts:612]: @guanbinrui this modify is safe?

File: packages/mask/src/plugins/Wallet/services/domain.ts (hunk 21 above)
Comment (guanbinrui) [reply packages/mask/src/plugins/Wallet/services/domain.ts:32]: I cannot accept this suggestion.

File: packages/web3-shared/evm/types/index.ts (hunk 22 above)
Comment (guanbinrui) [reply packages/web3-shared/evm/types/index.ts:612]: It's safe.

**Comment by wenluomask** [comment]
6. I put the RSS3 profile link to bio, it showed failed to find any address;
Screen Shot 2021-12-21 at 1 57 11 PM

7. I had NFT avatar, but didn't show my address in Web3;
Screen Shot 2021-12-21 at 2 13 50 PM

MingMing had NFT avatar and address in bio, but didn't show any address either;
Screen Shot 2021-12-21 at 2 14 25 PM

File: packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx [hunk 23]
```diff
@@ -232,9 +233,16 @@ export function CollectibleListAddress(props: CollectibleListAddressProps) {
...
 }
 
-export function CollectionList({ address }: { address: string }) {
+export function CollectionList({
+    address,
+    selectedNetwork,
+}: {
+    address: string
+    selectedNetwork: Web3Plugin.NetworkDescriptor | null
+}) {
```
Comment (septs) [inline packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx:242]: plz extract this type declare as named interface definition

**Comment by wenluomask** [comment]
Checked.

**Pull request merged**